### PR TITLE
Lms/admin diagnostic student report api

### PR DESCRIPTION
**Pull request by anathomical**
## WHAT
Continuing to split out code from a bigger branch into smaller bites.  Some of this code is already in #11588.  I'll make comments at the top of files where this is the case so that folks can review those files in just one place.

- Add a controller for Admin Diagnostic Student Report APIs
- - Fetching the student data from cache or kicking off a job to cache (following the pattern we've established elsewhere)
- - Fetching "filter scope" data so that we can report to an admin the size of the data-set that their filter selection should be showing them
- Add a worker to cache student report data on demand (again, following established patterns)
- Copy the base `DiagnosticAggregateViewQuery` from #11588 
- Fix a small bug in the `pre_post_diagnostic_skill_group_performance_view.sql` definition
- Write a query to get the majority of the student report info
- Write a query to get the specific recommendation information for the student report
- Write a query to handle "filter scope" numbers
## WHY
- Controller and worker pairing allows for on-demand cache generation and fetching report data from cache
- The queries provide all data needed to display on the front-end
## HOW
- Controller and worker are based on patterns established for BigQuery data fetching in Snapshots
- Major data query takes advantage of a materialized view to fetch all the data we need for the report
- Recommendation query fetches additional data that doesn't fit into the materialized view
- Filter scope calculates the size of the addressable student population for a given user and filter selection based on filters

### Notion Card Links
https://www.notion.so/quill/Admin-Diagnostic-Student-report-API-4190857056f64ebabeda69bd03000fb3?pvs=4

### What have you done to QA this feature?
(Provide enough detail that a reviewer could assess whether additional QA should be done. For larger projects, additionally use the Engineer Feature Testing Notion template. Review Guidelines if needed: https://www.notion.so/quill/Github-PR-QA-Guidelines-49e99fc965654ceeb8c6249bd9d181d7)

PR Checklist | Your Answer
------------ | -------------
Have you added and/or updated tests? |  Yes
Have you deployed to Staging? | Yes
Self-Review: Have you done an initial self-review of the code below on Github? | Yes

### CODE REVIEW

**Comment by happythenewsad** [comment]
> Continuing to split out code from a bigger branch into smaller bites. Some of this code is already in https://github.com/empirical-org/Empirical-Core/pull/11588. I'll make comments at the top of files where this is the case so that folks can review those files in just one place.

It may be faster to rebase this branch against 11588 after 11588 is merged...

File: services/QuillLMS/app/queries/admin_diagnostic_reports/diagnostic_aggregate_view_query.rb [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+    def school_ids_where_clause = "AND schools_users.school_id IN (#{school_ids.join(',')})"
+    def teacher_ids_where_clause = ("AND schools_users.user_id IN (#{teacher_ids.join(',')})" if teacher_ids.present?)
+
+    def group_by_clause = "GROUP BY performance.activity_id, activities.name, aggregate_id, #{aggregate_sort_clause}"
+
+    def active_classroom_stubs_view = materialized_view('active_classroom_stubs_view')
+    def active_user_names_view = materialized_view('active_user_names_view')
+    def performance_view = materialized_view('pre_post_diagnostic_skill_group_performance_view')
+
+    def materialized_views = [active_classroom_stubs_view, active_user_names_view, performance_view]
```
Comment (dandrabik) [inline services/QuillLMS/app/queries/admin_diagnostic_reports/diagnostic_aggregate_view_query.rb]: Let's move this to the top below initialize.

File: services/QuillLMS/app/queries/admin_diagnostic_reports/diagnostic_recommendations_by_student_query.rb [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+    def select_clause
+      <<-SQL
+        SELECT students.id AS student_id,
+          COUNT(DISTINCT activity_sessions.id) AS completed_activities,
+          SUM(activity_sessions.timespent) AS time_spent_seconds
+      SQL
+    end
+
+    def from_and_join_clauses
```
Comment (dandrabik) [inline services/QuillLMS/app/queries/admin_diagnostic_reports/diagnostic_recommendations_by_student_query.rb:30]: Per our discussion today, this is still a lot of joins outside of the view. I wonder if there is a way to make another view and join to that in a way that is more efficient. 

I think it's fine to merge this PR and leave that type of refactor for another, later PR. I just wanted to note it.

File: services/QuillLMS/app/queries/admin_diagnostic_reports/diagnostic_recommendations_by_student_query.rb [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+    def relevant_date_column = "activity_sessions.completed_at"
+    def relevant_diagnostic_where_clause = "AND recommendations.activity_id = #{diagnostic_id}"
+
+    def active_classroom_stubs_view = materialized_view('active_classroom_stubs_view')
+    def active_classroom_unit_stubs_view = materialized_view('active_classroom_unit_stubs_view')
+    def active_user_names_view = materialized_view('active_user_names_view')
+    def recommendation_activity_session_stubs_view = materialized_view('recommendation_activity_session_stubs_view')
+
+    def materialized_views = [recommendation_activity_session_stubs_view, active_classroom_stubs_view, active_classroom_unit_stubs_view, active_user_names_view]
```
Comment (dandrabik) [inline services/QuillLMS/app/queries/admin_diagnostic_reports/diagnostic_recommendations_by_student_query.rb]: Move this up (I won't comment on the rest of them).

File: services/QuillLMS/app/queries/admin_diagnostic_reports/diagnostic_recommendations_by_student_query.rb [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+      SQL
+    end
+
+    private def post_query_transform(results)
+      results.to_h do |result|
+        [result[:student_id], {completed_activities: result[:completed_activities], time_spent_seconds: result[:time_spent_seconds]}]
+      end
+    end
+
+    private def valid_aggregation_options
```
Comment (dandrabik) [inline services/QuillLMS/app/queries/admin_diagnostic_reports/diagnostic_recommendations_by_student_query.rb]: These two methods could be one liners

File: services/QuillLMS/lib/query_examples/admin_diagnostics/diagnostic-skills.sql [hunk 5]
```diff
@@ -0,0 +1,176 @@
+        /*
```
Comment (dandrabik) [inline services/QuillLMS/lib/query_examples/admin_diagnostics/diagnostic-skills.sql:1]: If I'm reading this correctly, this view is no longer used, correct?. That's would have been a huge data bill and is 18 seconds.

File: services/QuillLMS/lib/query_examples/admin_diagnostics/diagnostic-students-view.sql [hunk 6]
```diff
@@ -0,0 +1,82 @@
+        /*
+           Data Processed By Query: 1.5 GB
+           Bytes Billed For Query:  0.0 GB
+           Total Query Time:        7521 ms
```
Comment (dandrabik) [inline services/QuillLMS/lib/query_examples/admin_diagnostics/diagnostic-students-view.sql]: Yeah, we likely have to do another pass on this one. 7.5 seconds is too long. Could be in a different PR.

**Review by dandrabik — APPROVED** [review]
Looks good for merging (small comments about file organization). There are some performance items that I think we want to revisit, but it's probably better to merge and focus on those in another PR.

File: services/QuillLMS/lib/query_examples/admin_diagnostics/diagnostic-skills.sql (hunk 5 above)
Comment (anathomical) [reply services/QuillLMS/lib/query_examples/admin_diagnostics/diagnostic-skills.sql:1]: Correct, the non-view version is completely replaced by the new view-based version.  I included this as a comparison.

File: services/QuillLMS/lib/query_examples/admin_diagnostics/diagnostic-students-view.sql (hunk 6 above)
Comment (anathomical) [reply services/QuillLMS/lib/query_examples/admin_diagnostics/diagnostic-students-view.sql]: I reran the thor script this morning, and this time it came through at 4030ms, which is probably closer to acceptable.  I agree that we should do a more robust set of benchmarking to understand average and max run times, and that this will likely indicate a performance pass.

File: services/QuillLMS/lib/query_examples/admin_diagnostics/diagnostic-students-view.sql [hunk 7]
```diff
@@ -0,0 +1,82 @@
+        /*
+           Data Processed By Query: 1.5 GB
+           Bytes Billed For Query:  0.0 GB
+           Total Query Time:        4030 ms
```
Comment (anathomical) [inline services/QuillLMS/lib/query_examples/admin_diagnostics/diagnostic-students-view.sql]: Note: a previous run of this benchmarking script (with no code changes), ran at `7521ms`.  We'll want to do a more robust examination of actual performance here to understand average and max execution times.

File: services/QuillLMS/app/queries/admin_diagnostic_reports/student_count_by_filter_scope_query.rb [hunk 8]
```diff
@@ -0,0 +1,81 @@
+# frozen_string_literal: true
+
+module AdminDiagnosticReports
+  class StudentCountByFilterScopeQuery < ::QuillBigQuery::MaterializedViewQuery
+    class InvalidDiagnosticIdError < StandardError; end
+
+    attr_reader :timeframe_start, :timeframe_end, :school_ids, :grades, :teacher_ids, :classroom_ids, :user, :additional_aggregation, :diagnostic_id
+
+    DIAGNOSTIC_ORDER_BY_ID = [
```
Comment (happythenewsad) [inline services/QuillLMS/app/queries/admin_diagnostic_reports/student_count_by_filter_scope_query.rb:9]: This list can be accessed from DiagnosticAggregateQuery, so we don't need to reimplement it here

**Review by happythenewsad — APPROVED** [review]
Lots of code; clearly written

LGTM

I added a comment about a double-implementation of the `DIAGNOSTIC_ORDER_BY_ID` lookup table that we may want to DRY up at some point
